R/utils.R

```diff
@@ -81,12 +81,30 @@ convertWildCards <- function(val) {
   gsub("_", "(.)", val)
 }
 
-getSelector <- function(key, value) {
+getSelector <- function(key, value, exact) {
+  if ( is.null(value ) ) {
+    return("")
+  }
   comp <- " = "
-  if (isPatternValue(value)) {
+  if (( value == "%" || !exact ) &&
+      isPatternValue(value)) {
     comp <- " LIKE "
   }
+  if (exact && (value != "%"))
+    value <- escapeChars(value)
   value <- paste0("'", value, "'")
 
   paste0(" AND ", key, comp, value)
 }
+
+# Will iterate over charsToEsc argument and for each:
+# will escape any un-escaped occurance in `x`.
+escapeChars <- function(x, charsToEsc = c("_"), escChar ="\\\\") {
```

Comment:
I think it would be better to call this `escapePattern()` and hard code it to escape `%` and `_`.

Reply:
Hey Hadley:

Renamed it - but I am not sold on escaping "%".  It seems to me that it is *so* infrequently used as an identifier that if someone is passing "%" sign they are *trying* to use a wild card ( but may have forgotten to wrap the input appropriately ).

I keep thinking, should we escape the percent character in `odbcConnectionColumns(conn, catalog_name = "my_cat", schema_name = "my_sch", name = "my_tbl", column_name = "%", exact = TRUE)`?

Anyway, I can be moved here just let me know.

Reply:
My feeling is that we want to make the leaky abstraction on top of the ODBC API clear, so there are two choices, either a pattern match (the default) or a literal string. 

I think what you're recognising in the API is that it's a bit weird to have one parameter that controls whether or not three other arguments are escaped. That's one advantage of the interface that uses `I()`, but that would mean making `escape = TRUE` the default, which I know you're not keen on. 

So only escaping `_` seems like a pragmatic choice and I'm fine with it if you want to revert back to your original idea.

Reply:
Hey - thanks for continuing to think about this  - good points about pros/cons.  I am still thinking about how the user would pass `I()` through the `DBI::Id` container.

```
  > str(DBI::Id(name=I("abc"))@name)
   Named chr "abc"
   - attr(*, "names")= chr "name"
```

In the meantime - I think by escaping the identifiers in the `DBI` API endpoints we've covered most (all?) of the performance gains to be had.  Copy your note on let's merge and then iterate as needed.